csrc/device_lower/pass/alias_memory.cpp

```diff
@@ -1382,6 +1383,50 @@ class AllocationReuseModifier : private kir::ExprMutator {
   std::unordered_map<kir::Allocate*, kir::Allocate*> old2new_;
 };
 
+//! Set addresses without any re-use of shared memory. This simply scans
+//! through exprs and each time we find an unaliased Allocate align the current
+//! address and increment it by the given amount.
+class NoReuseSharedMemAllocator : kir::IrVisitor {
+ public:
+  void allocate(std::vector<Expr*>& exprs) {
+    handle(exprs);
+  }
+
+ private:
+  using kir::IrVisitor::handle;
+
+  static Val* alignExpr(Val* addr, int64_t alignment = 16) {
+    if (alignment == 1) {
+      return addr;
+    }
+    auto n_minus_one = IrBuilder::create<Val>(alignment - 1);
+    return bitwise_and(add(addr, n_minus_one), bitwise_not(n_minus_one));
+  }
+
+  void handle(kir::Allocate* alloc) final {
+    if (alloc->memoryType() != MemoryType::Shared || alloc->alias()) {
+      return;
+    }
+
+    const auto buffer_dtype = alloc->buffer()->dtype();
+    const auto dtype_size = dataTypeSize(buffer_dtype);
+
+    auto address = current_offset ? alignExpr(current_offset, 16 /*dtype_size*/)
+                                  : FusionGuard::getCurFusion()->zeroVal();
+
+    alloc->setAddress(address);
+
+    auto size_bytes = dtype_size == 1
+        ? alloc->size()
+        : mul(alloc->size(), IrBuilder::create<Val>(dtype_size));
+
+    current_offset = add(address, size_bytes);
+  }
+
+ private:
+  Val* current_offset = nullptr;
+};
+
 } // namespace
 
 std::vector<Expr*> reuseMemoryAllocations(const std::vector<Expr*>& exprs) {
```

Comment:
Is it possible to move the `reuseMemoryAllocations` down after `commonScalarMap().initialize(exprs_unrolled_loops);` so that we can do index hoisting? (If the answer is yes, then let's do it in a separate PR, don't change anything here.)

Reply:
Yeah I was wondering about this; since the expressions are just cumulative sums they have tons of common subexpressions. But wouldn't moving it down later mean we could not hoist these common expressions since they would not yet be defined?

Reply:
> But wouldn't moving it down later mean we could not hoist these common expressions since they would not yet be defined?

I don't understand. What do you mean?

`commonScalarMap().initialize(exprs_unrolled_loops);` just initializes the data structure for index hoisting and figure out what scalars are already computed as part of fusion definition. Once this data structure is initialized, we can use `GpuLower::current()->commonScalarMap().hoistScalar(val, for_loops_);` and common subexpr elimination should be automatically handled.

Reply:
We can change `alloc->setAddress(address);` into
```C++
alloc->setAddress(GpuLower::current()->commonScalarMap().hoistScalar(address, for_loop_));
```